source/agora/node/Ledger.d

```diff
@@ -98,54 +106,154 @@ public class Ledger
+
+    private void addValidatedBlock (const ref Block block) nothrow @safe
+    {
+        // add the new UTXOs
+        block.txs.each!(tx => this.utxo_set.updateUtxoCache(tx));
+
+        // remove the TXs from the Pool
+        block.txs.each!(
+            (tx)
+            {
+                try
+                {
+                    this.pool.remove(tx.hashFull());
+                }
+                catch (Exception ex)
+                {
+                    // this OP must not fail
+                    assert(0);
+                }
+            }
+        );
+
+        if (!this.storage.saveBlock(block))
+            assert(0);
+
+        // read back and cache the last block
+        if (!this.storage.readLastBlock(this.last_block))
+            assert(0);
+    }
+
+    /***************************************************************************
+
+        Try making a new block if there are enough valid and non double-spending
+        transactions in the pool
+
+        Double-spending transactions will be skipped over while iterating
+        over the pool. If there are not enough valid transactions,
+        a block will not be created.
 
     ***************************************************************************/
 
-    private void makeBlock () @safe
+    private void tryCreateBlock () @safe
     {
         Hash[] hashes;
         Transaction[] txs;
 
+        this.utxo_set.reset();
         foreach (hash, tx; this.pool)
         {
-            hashes ~= hash;
-            txs ~= tx;
+            if (tx.isValid(&this.utxo_set.findUTXO))
+            {
+                hashes ~= hash;
+                txs ~= tx;
+            }
+            else
+            {
+                logDebug("Rejected double-spend tx: %s", tx);
+            }
+
+            if (txs.length >= Block.TxsInBlock)
+                break;
         }
 
-        assert(txs.length == Block.TxsInBlock);
+        if (txs.length != Block.TxsInBlock)
+            return;  // not enough valid txs
 
         auto block = makeNewBlock(this.last_block, txs);
-        if (!this.acceptBlock(block))
+        if (!this.acceptBlock(block))  // txs should be valid
             assert(0);
+    }
+
+    /***************************************************************************
+
+        Check whether the transaction is valid and may be added to the pool.
+
+        A transaction is valid if it references a previous UTXO in the
+        blockchain. Note that double-spend transactions are not tracked here,
+        they are only tracked during the creation of a block.
+
+        Params:
+            transaction = the transaction to validate
+
+        Returns:
+            true if the transaction may be added to the pool
+
+    ***************************************************************************/
+
+    public bool isValidTransaction (const ref Transaction tx) nothrow @safe
+    {
+        this.utxo_set.reset();
```

Comment:
This is quite dangerous, but okay for now. Maybe open an issue about it ?

Reply:
Badly named method, I'll rename it.